Unreal Marketplace Testing Failure for Android for v1.3.0

For v1.3.0, we added Android to the whitelist platform in the uplugin file. This is causing them to run into C++17 compilation issues similar to https://github.com/CesiumGS/cesium-unreal/issues/190#issuecomment-831522727.

[Cesium Errors.txt](https://github.com/CesiumGS/cesium-unreal/files/6579898/Cesium.Errors.txt)

If we don't need Android support through marketplace for this release, they are recommending that we submit the plugin by removing `Android` from the uplugin Supported and Target platforms, and instead move it to `BlacklistPlatforms` for this release.

Regarding the errors: As mentioned in the issue comment above, I did see them when trying to build the plugin for Android at one point, but they essentially disappeared one day. I'm still not sure why they don't show up on CI. Those errors also popped up when I was trying to compile Android with UE5 EA over the weekend.

CC @CesiumGS/unreal

@shehzan10 I am attempting to locally package the plugin for Android on the current `main` branch. However I am running into the mikktspace dependency error that was discussed in #444

Beforehand, I compiled the extern directory with VS 2017, and then ran the `Windows + Android` example given in the [packaging documentation](https://github.com/CesiumGS/cesium-unreal/blob/main/Documentation/packaging-guide.md)

Got the following error:
```
  Reflection code generated for UE4Game in 2.5432914 seconds
  Compiling Native 32-bit code with NDK API 'android-21'
  Compiling Native 64-bit code with NDK API 'android-21'
  Writing manifest to C:\Users\Xuelong\Dev\CesiumDevTest\Packages\HostProject\Saved\Manifest-UE4Game-Android-Development.xml
  Building UE4Game...
  Building 6 actions with 12 processes...
    [1/6] PCH.CesiumRuntime.h [armv7]
    [2/6] PCH.CesiumRuntime.h [arm64]
    [3/6] Module.CesiumRuntime.gen.cpp [armv7]
    [4/6] Module.CesiumRuntime.gen.cpp [arm64]
    [5/6] Module.CesiumRuntime.cpp [armv7]
    In file included from C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Intermediate/Build/Android/UE4/Development/CesiumRuntime/Module.CesiumRuntime.cpp:12:
    C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumGltfComponent.cpp(37,10): fatal error: 'mikktspace.h' file not found
    #include "mikktspace.h"
             ^~~~~~~~~~~~~~
    1 error generated.
    [6/6] Module.CesiumRuntime.cpp [arm64]
    In file included from C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Intermediate/Build/Android/UE4/Development/CesiumRuntime/Module.CesiumRuntime.cpp:12:
    C:/Users/Xuelong/Dev/CesiumDevTest/Packages/HostProject/Plugins/CesiumForUnreal/Source/CesiumRuntime/Private/CesiumGltfComponent.cpp(37,10): fatal error: 'mikktspace.h' file not found
    #include "mikktspace.h"
             ^~~~~~~~~~~~~~
    1 error generated.
Took 19.8765484s to run UnrealBuildTool.exe, ExitCode=6
UnrealBuildTool failed. See log for more details. (C:\Users\Xuelong\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Game-Android-Development.txt)
AutomationTool exiting with ExitCode=6 (6)
BUILD FAILED
```

Making the plugin files readonly after extracting them does cause the build to fail, but in a different and fairly predictable way:

```
UATHelper: Package Plugin Task (Windows): Running AutomationTool...
UATHelper: Package Plugin Task (Windows): Parsing command line: BuildPlugin -Plugin=C:/Build/++Portal/Sync/LocalBuilds/PluginTemp/HostProject/Plugins/CesiumForUnreal/CesiumForUnreal.uplugin -Package=C:/Build/++Portal/Sync/LocalBuilds/PluginTemp/package/CesiumForUnreal -CreateSubFolder
UATHelper: Package Plugin Task (Windows): Copying 1242 file(s) using max 64 thread(s)
UATHelper: Package Plugin Task (Windows): Reading plugin from C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\CesiumForUnreal.uplugin...
UATHelper: Package Plugin Task (Windows): Building plugin for host platforms: Win64
UATHelper: Package Plugin Task (Windows): Running: C:\Program Files\Epic Games\UE_4.26\Engine\Binaries\DotNET\UnrealBuildTool.exe UE4Editor Win64 Development -Project=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\HostProject.uproject -plugin=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\
Plugins\CesiumForUnreal\CesiumForUnreal.uplugin -iwyu -noubtmakefiles -manifest=C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Saved\Manifest-UE4Editor-Win64-Development.xml -nohotreload -log="C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Editor-Win64-Development.
txt"
UATHelper: Package Plugin Task (Windows):   Parsing headers for UE4Editor
UATHelper: Package Plugin Task (Windows):     Running UnrealHeaderTool "C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\HostProject.uproject" "C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Intermediate\Build\Win64\UE4Editor\Development\UE4Editor.uhtmanifest" -LogCmds="loginit warning, loge
xit warning, logdatabase error" -Unattended -WarningsAsErrors -abslog="C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UHT-UE4Editor-Win64-Development.txt" -installed
UATHelper: Package Plugin Task (Windows):   Reflection code generated for UE4Editor in 8.242515 seconds
UATHelper: Package Plugin Task (Windows):   Writing manifest to C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Saved\Manifest-UE4Editor-Win64-Development.xml
UATHelper: Package Plugin Task (Windows):   Building UE4Editor...
UATHelper: Package Plugin Task (Windows):   Using Visual Studio 2019 14.29.30038 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30037) and Windows 10.0.19041.0 SDK (C:\Program Files (x86)\Windows Kits\10).
UATHelper: Package Plugin Task (Windows):   Building 13 actions with 12 processes...
UATHelper: Package Plugin Task (Windows):     [1/13] Default.rc2
UATHelper: Package Plugin Task (Windows):     [2/13] Default.rc2
UATHelper: Package Plugin Task (Windows):     [3/13] PCH.CesiumRuntime.cpp
UATHelper: Package Plugin Task (Windows):     [4/13] PCH.CesiumEditor.cpp
UATHelper: Package Plugin Task (Windows):     [5/13] Module.CesiumEditor.gen.cpp
UATHelper: Package Plugin Task (Windows):     [6/13] Module.CesiumRuntime.gen.cpp
UATHelper: Package Plugin Task (Windows):     [7/13] Module.CesiumRuntime.cpp
UATHelper: Package Plugin Task (Windows):     [8/13] UE4Editor-CesiumRuntime.lib
UATHelper: Package Plugin Task (Windows):        Creating library C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumRuntime\UE4Editor-CesiumRuntime.lib and object C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugin
s\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumRuntime\UE4Editor-CesiumRuntime.exp
UATHelper: Package Plugin Task (Windows):     [9/13] UE4Editor-CesiumRuntime.dll
UATHelper: Package Plugin Task (Windows):     LINK : fatal error LNK1104: cannot open file 'C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Binaries\Win64\UE4Editor-CesiumRuntime.dll'
UATHelper: Package Plugin Task (Windows):     [10/13] Module.CesiumEditor.cpp
UATHelper: Package Plugin Task (Windows):     [11/13] UE4Editor-CesiumEditor.lib
UATHelper: Package Plugin Task (Windows):        Creating library C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumEditor\UE4Editor-CesiumEditor.lib and object C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\
CesiumForUnreal\Intermediate\Build\Win64\UE4Editor\Development\CesiumEditor\UE4Editor-CesiumEditor.exp
UATHelper: Package Plugin Task (Windows):     [12/13] UE4Editor-CesiumEditor.dll
UATHelper: Package Plugin Task (Windows):     LINK : fatal error LNK1104: cannot open file 'C:\Build\++Portal\Sync\LocalBuilds\PluginTemp\package\CesiumForUnreal\HostProject\Plugins\CesiumForUnreal\Binaries\Win64\UE4Editor-CesiumEditor.dll'
UATHelper: Package Plugin Task (Windows): Took 53.875737s to run UnrealBuildTool.exe, ExitCode=6
UATHelper: Package Plugin Task (Windows): UnrealBuildTool failed. See log for more details. (C:\Users\kring\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Editor-Win64-Development.txt)
UATHelper: Package Plugin Task (Windows): AutomationTool exiting with ExitCode=6 (6)
UATHelper: Package Plugin Task (Windows): BUILD FAILED
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -12,13 +12,8 @@ jobs:
     - npm install
     - npm run format -- --dry-run -Werror
   - stage: Build OS Packages
-    name: Windows + VS2017
+    name: Windows
     os: windows
-    cache:
-      timeout: 1000
-      directories:
-      - /C/Program\ Files/Epic\ Games/UE_4.26
-      - /C/Program\ Files\ \(x86\)/Epic\ Games/Launcher
     install:
     - choco install 7zip.portable
     - choco install python --version 3.9.2
@@ -51,8 +46,6 @@ jobs:
     cache:
       timeout: 1000
       directories:
-      - /C/Program\ Files/Epic\ Games/UE_4.26
-      - /C/Program\ Files\ \(x86\)/Epic\ Games/Launcher
       - $HOME/android-ndk-r21e
     install:
     - choco install 7zip.portable
@@ -70,24 +63,18 @@ jobs:
     - export NDKROOT=$ANDROID_NDK_ROOT
     - export CESIUM_UNREAL_VERSION=$(git describe)
     - export BUILD_CESIUM_UNREAL_PACKAGE_NAME=CesiumForUnreal-android-${CESIUM_UNREAL_VERSION}
-    - mkdir -p extern/build-win64
     - mkdir -p extern/build-android
-    - cd extern/build-win64
-    - cmake .. -A x64
-    - cd ../../extern/build-android
+    - cd extern/build-android
     - cmake .. -G Ninja -DCMAKE_TOOLCHAIN_FILE="unreal-android-toolchain.cmake" -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DCMAKE_BUILD_TYPE=Release
     - cd ../..
     script:
-    - cd extern/build-win64
-    - cmake --build . --config Release --target install -- /maxcpucount:4
-    - cd ../../extern/build-android
+    - cd extern/build-android
     - cmake --build . --config Release --target install
     - cd ../..
     - rm -rf extern
     - export CLONEDIR=$PWD
     - cd /c/Program\ Files/Epic\ Games/UE_4.26/Engine/Build/BatchFiles
-    - ./RunUAT.bat BuildPlugin -Plugin="$CLONEDIR/CesiumForUnreal.uplugin" -Package="$CLONEDIR/../packages/CesiumForUnreal" -CreateSubFolder -TargetPlatforms=Win64+Android
-    - cat "C:\Users\travis\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_4.26\UBT-UE4Game-Android-Development.txt"
+    - ./RunUAT.bat BuildPlugin -Plugin="$CLONEDIR/CesiumForUnreal.uplugin" -Package="$CLONEDIR/../packages/CesiumForUnreal" -CreateSubFolder -TargetPlatforms=Android -NoHostPlatform
     - cd "$CLONEDIR/../packages"
     - 7z a ${BUILD_CESIUM_UNREAL_PACKAGE_NAME}.zip CesiumForUnreal/
     - aws s3 cp ${BUILD_CESIUM_UNREAL_PACKAGE_NAME}.zip s3://builds-cesium-unreal/

--- a/Source/CesiumEditor/CesiumEditor.Build.cs
+++ b/Source/CesiumEditor/CesiumEditor.Build.cs
@@ -5,6 +5,7 @@ using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Reflection;
 
 public class CesiumEditor : ModuleRules
 {
@@ -132,5 +133,24 @@ public class CesiumEditor : ModuleRules
         PCHUsage = PCHUsageMode.UseExplicitOrSharedPCHs;
         PrivatePCHHeaderFile = "Private/PCH.h";
         CppStandard = CppStandardVersion.Cpp17;
+
+        if (Target.Platform == UnrealTargetPlatform.Android &&
+            Target.Version.MajorVersion == 4 &&
+            Target.Version.MinorVersion == 26 &&
+            Target.Version.PatchVersion < 2)
+        {
+            // In UE versions prior to 4.26.2, the Unreal Build Tool on Android
+            // (AndroidToolChain.cs) ignores the CppStandard property and just
+            // always uses C++14. Our plugin can't be compiled with C++14.
+            //
+            // So this hack uses reflection to add an additional argument to
+            // the compiler command-line to force C++17 mode. Clang ignores all
+            // but the last `-std=` argument, so the `-std=c++14` added by the
+            // UBT is ignored.
+            Type type = Target.GetType();
+            FieldInfo innerField = type.GetField("Inner", BindingFlags.Instance | BindingFlags.NonPublic);
+            TargetRules inner = (TargetRules)innerField.GetValue(Target);
+            inner.AdditionalCompilerArguments += " -std=c++17";
+        }
     }
 }

--- a/Source/CesiumRuntime/CesiumRuntime.Build.cs
+++ b/Source/CesiumRuntime/CesiumRuntime.Build.cs
@@ -5,6 +5,7 @@ using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Reflection;
 
 public class CesiumRuntime : ModuleRules
 {
@@ -127,7 +128,7 @@ public class CesiumRuntime : ModuleRules
                 "Projects"
             }
         );
-        
+
         // Use UE's MikkTSpace on non-Android
         if (Target.Platform != UnrealTargetPlatform.Android)
         {
@@ -176,5 +177,19 @@ public class CesiumRuntime : ModuleRules
         PrivatePCHHeaderFile = "Private/PCH.h";
         CppStandard = CppStandardVersion.Cpp17;
         bEnableExceptions = true;
-    }
+
+        if (Target.Platform == UnrealTargetPlatform.Android)
+        {
+          // This line is just to force C++14 even on UE 4.262, so we can test what follows.
+          CppStandard = CppStandardVersion.Cpp14;
+
+          // Use reflection to force C++17 mode even though the UBT in UE
+          // versions prior to 4.26.2 ignore the CppStandard property
+          // and just always use C++14.
+          Type type = Target.GetType();
+          FieldInfo innerField = type.GetField("Inner", BindingFlags.Instance | BindingFlags.NonPublic);
+          TargetRules inner = (TargetRules)innerField.GetValue(Target);
+          inner.AdditionalCompilerArguments += " -std=c++17";
+        }
+  }
 }

--- a/Source/CesiumRuntime/CesiumRuntime.Build.cs
+++ b/Source/CesiumRuntime/CesiumRuntime.Build.cs
@@ -178,18 +178,23 @@ public class CesiumRuntime : ModuleRules
         CppStandard = CppStandardVersion.Cpp17;
         bEnableExceptions = true;
 
-        if (Target.Platform == UnrealTargetPlatform.Android)
+        if (Target.Platform == UnrealTargetPlatform.Android &&
+            Target.Version.MajorVersion == 4 &&
+            Target.Version.MinorVersion == 26 &&
+            Target.Version.PatchVersion < 2)
         {
-          // This line is just to force C++14 even on UE 4.262, so we can test what follows.
-          CppStandard = CppStandardVersion.Cpp14;
-
-          // Use reflection to force C++17 mode even though the UBT in UE
-          // versions prior to 4.26.2 ignore the CppStandard property
-          // and just always use C++14.
-          Type type = Target.GetType();
-          FieldInfo innerField = type.GetField("Inner", BindingFlags.Instance | BindingFlags.NonPublic);
-          TargetRules inner = (TargetRules)innerField.GetValue(Target);
-          inner.AdditionalCompilerArguments += " -std=c++17";
+            // In UE versions prior to 4.26.2, the Unreal Build Tool on Android
+            // (AndroidToolChain.cs) ignores the CppStandard property and just
+            // always uses C++14. Our plugin can't be compiled with C++14.
+            //
+            // So this hack uses reflection to add an additional argument to
+            // the compiler command-line to force C++17 mode. Clang ignores all
+            // but the last `-std=` argument, so the `-std=c++14` added by the
+            // UBT is ignored.
+            Type type = Target.GetType();
+            FieldInfo innerField = type.GetField("Inner", BindingFlags.Instance | BindingFlags.NonPublic);
+            TargetRules inner = (TargetRules)innerField.GetValue(Target);
+            inner.AdditionalCompilerArguments += " -std=c++17";
         }
-  }
+    }
 }

--- a/travis/travis-get-ue.sh
+++ b/travis/travis-get-ue.sh
@@ -1,16 +1,12 @@
-if [[ $TRAVIS_OS_NAME == "windows" && ! -d "/c/Program Files/Epic Games/UE_4.26" ]]
+if [[ $TRAVIS_OS_NAME == "windows" ]]
 then
     # Enable compression because disk space is limited on Travis.
     mkdir "C:\Program Files\Epic Games"
-    compact //c "//s:C:\Program Files\Epic Games"
-    mkdir "C:\Program Files (x86)\Epic Games"
-    compact //c "//s:C:\Program Files (x86)\Epic Games"
-    AWS_ACCESS_KEY_ID=${GOOGLE_ACCESS_KEY_ID} AWS_SECRET_ACCESS_KEY=${GOOGLE_SECRET_ACCESS_KEY} aws s3 --endpoint-url https://storage.googleapis.com cp s3://cesium-unreal-engine/windows-2021-05-19/UE_4.26.zip .
-    7z x UE_4.26.zip "-oC:\Program Files\Epic Games"
-    rm UE_4.26.zip
-    AWS_ACCESS_KEY_ID=${GOOGLE_ACCESS_KEY_ID} AWS_SECRET_ACCESS_KEY=${GOOGLE_SECRET_ACCESS_KEY} aws s3 --endpoint-url https://storage.googleapis.com cp s3://cesium-unreal-engine/windows-2021-05-07/Launcher.zip .
-    7z x Launcher.zip "-oC:\Program Files (x86)\Epic Games"
-    rm Launcher.zip
+    mkdir "C:\Program Files\Epic Games\UE_4.26"
+    compact //c "//s:C:\Program Files\Epic Games\UE_4.26"
+    AWS_ACCESS_KEY_ID=${GOOGLE_ACCESS_KEY_ID} AWS_SECRET_ACCESS_KEY=${GOOGLE_SECRET_ACCESS_KEY} aws s3 --endpoint-url https://storage.googleapis.com cp s3://cesium-unreal-engine/windows-2021-07-12/UnrealEngine-4-26-0-minimal.zip .
+    7z x UnrealEngine-4-26-0-minimal.zip "-oC:\Program Files\Epic Games\UE_4.26"
+    rm UnrealEngine-4-26-0-minimal.zip
 elif [[ $TRAVIS_OS_NAME == "osx" ]]
 then
     aws s3 cp s3://cesium-unreal-engine/macos-2021-04-13/UE_4.26.tar.gz .

